### PR TITLE
Expose Jenkins logs from jenkins-operator

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/kubernetes/test-infra/issues/3407

/cc @spxtr @cjwagner

### CODE REVIEW

**Comment by kargakis** [comment]
/hold

Seems like url parameters are not forwarded by kube-proxy?
```
$ curl http://deck/log?job=test_branch_origin_extended_networking&id=887                                                                                                             
[1] 64
bash-4.2$ Missing job and ID query
```

**Comment by kargakis** [comment]
deck will use a url template for Jenkins so it can also not use url parameters at all.

**Comment by kargakis** [comment]
Tested in our cluster and it works like a charm.

File: prow/cmd/deck/jobs.go [hunk 1]
```diff
@@ -114,11 +116,22 @@ func (ja *JobAgent) GetJobLog(job, id string) ([]byte, error) {
...
-	num, err := strconv.Atoi(id)
-	if err != nil {
-		return nil, err
+	for _, agentToTmpl := range ja.c.Config().Deck.ExternalAgentLogs {
+		if agentToTmpl.Agent != string(j.Spec.Agent) {
+			continue
+		}
+		var b bytes.Buffer
+		if err := agentToTmpl.URLTemplate.Execute(&b, &j); err != nil {
+			return nil, fmt.Errorf("error executing URL template: %v", err)
```
Comment (cjwagner) [inline prow/cmd/deck/jobs.go]: Include the agent and prowjob in the error message like you do bellow.

File: prow/cmd/jenkins-operator/logs.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+func handleLog(jc *jenkins.Client) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Cache-Control", "no-cache")
+		w.Header().Set("Access-Control-Allow-Origin", "*")
+		w.Header().Set("Access-Control-Allow-Methods", "GET")
+
+		// Needs to be a GET request.
+		if r.Method != http.MethodGet {
+			http.Error(w, "Forbidden.", http.StatusForbidden)
```
Comment (cjwagner) [inline prow/cmd/jenkins-operator/logs.go]: Use 405 (Method not allowed).

File: prow/cmd/jenkins-operator/logs.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		// Needs to get Jenkins logs.
+		if !strings.HasSuffix(r.URL.Path, "consoleText") {
+			http.Error(w, "Forbidden.", http.StatusForbidden)
+			return
+		}
+
+		log, err := jc.Get(r.URL.Path)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Log not found: %v", err), http.StatusNotFound)
+			logrus.WithError(err).Warning("Error returned.")
```
Comment (cjwagner) [inline prow/cmd/jenkins-operator/logs.go]: Please include more context in this error message. Specifically, that the error is a result of trying to fetch logs from jenkins.

File: prow/cmd/deck/jobs.go (hunk 1 above)
Comment (kargakis) [reply prow/cmd/deck/jobs.go]: done

File: prow/cmd/jenkins-operator/logs.go (hunk 2 above)
Comment (kargakis) [reply prow/cmd/jenkins-operator/logs.go]: done

File: prow/cmd/jenkins-operator/logs.go (hunk 3 above)
Comment (kargakis) [reply prow/cmd/jenkins-operator/logs.go]: done

**Pull request merged**